### DISCUSSION TITLE
FIDO2/U2F/WebAuthn

### DISCUSSION THREAD

**Discussion by Beanow** [tauri-apps/tauri#6601] [General]
So the landscape and story surrounding Webauthn and friends is complicated and in flux.
Recently FIDO2 and some passkey announcements look to be shaking up this space this year and beyond.
https://fidoalliance.org/passkeys/

Unfortunately however Webview based applications seem to be in the *worst* shape of all options right now. Meaning you will have to either use a supported Browser, or use native OS libraries (when they exist). And we're not alone in this as you'd expect: https://github.com/electron/electron/issues/24573

Android SDK: [Fido2ApiClient](https://developers.google.com/android/reference/com/google/android/gms/fido/fido2/Fido2ApiClient)
    Windows Hello: [FIDO2 Win32 APIs](https://github.com/Microsoft/webauthn)
    macOS, iOS: WebAuthn via [ASWebAuthenticationSession](https://developer.apple.com/documentation/authenticationservices/aswebauthenticationsession)
Linux: :shrug: :sweat_smile: 

(I haven't gotten the current listing of all Webview opinions on this matter.)

Right now, for Tauri users this means they don't have reliable cross-platform access to these features in the Webview, using [Webauthn](https://caniuse.com/?search=webauthn). And the alternative is to use the native calls, which also aren't on every platform.

Now I don't want to get too anti here, but there's something more to keep in mind about the "native" libraries listed there. Notice that standards are available at every level:
- [W3C Webauthn](https://w3c.github.io/webauthn/)
- [CTAP 2](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html)
- USB HID / BLE / NFC
- [FIDO 2 / FIDO U2F](https://fidoalliance.org/specifications/download/)

Yet where do these native libraries live?
- Google Play Services.
- Windows Hello.
- iCloud keychains (ok tbh I'm not a mac expert so correct me here.)

Unfortunately, the big platforms seems to have rolled their own implementation and laced it *with an agenda* to use their accounts and services. :cry: There are positives to it as well though. Such as syncing passkeys within that ecosystem. But given how many standards are available. There seems to be a distinct lack of a vendor neutral implementation here.

Enter cool initiatives like https://github.com/AlfioEmanueleFresta/xdg-credentials-portal
An attempt to create the missing Linux portal for it, along with a rust implementation while they're at it :open_mouth:
Previously based off https://github.com/mozilla/authenticator-rs which is a WIP crate to add FIDO2 support to Firefox.

And another vendor shows here https://hwsecurity.dev/guide/fido-webview/ it is feasible to independently implement this and bridge it into a webview. (Java implemenation for Android / Xamarin there)

Something I've not extensively looked at yet, is any official stance / status on *every* webview.

Mentions:
https://github.com/tauri-apps/tauri/discussions/2261
https://github.com/tauri-apps/tauri/issues/6471
https://discord.com/channels/616186924390023171/1090685771754766481

**Comment by Beanow** [comment]
Note we also have:
https://github.com/tauri-apps/plugins-workspace/tree/dev/plugins/authenticator

**Reply by DeepBhat** [reply]
That link is broken can you please provide the correct one?

**Reply by FabianLars** [reply]
For tauri v1: https://github.com/tauri-apps/plugins-workspace/tree/v1/plugins/authenticator

For tauri v2: https://github.com/tauri-apps/plugins-workspace/tree/v2/plugins/authenticator

Note that we currently consider this plugin basically broken. It can serve as inspiration, but probably shouldn't be used as-is.

**Reply by dev-itzarun** [reply]
is there anyother available ?
